modal/_container_entrypoint.py

```diff
@@ -122,102 +171,12 @@ def _sigint_handler():
-                message = api_pb2.GeneratorDone(items_total=item_count)
-                container_io_manager.push_output(input_id, started_at, message, api_pb2.DATA_FORMAT_GENERATOR_DONE)
-            else:
-                if inspect.iscoroutine(res) or inspect.isgenerator(res) or inspect.isasyncgen(res):
-                    raise InvalidError(
-                        f"Sync (non-generator) function return value of type {type(res)}."
-                        " You might need to use @app.function(..., is_generator=True)."
-                    )
-                container_io_manager.push_output(input_id, started_at, res, imp_fun.data_format)
-        reset_context()
-
-    if imp_fun.input_concurrency > 1:
-        # We can't use `concurrent.futures.ThreadPoolExecutor` here because in Python 3.11+, this
-        # class has no workaround that allows us to exit the Python interpreter process without
-        # waiting for the worker threads to finish. We need this behavior on SIGINT.
-
-        import queue
-        import threading
-
-        spawned_workers = 0
-        inputs: queue.Queue[Any] = queue.Queue()
-        finished = threading.Event()
-
-        def worker_thread():
-            while not finished.is_set():
-                try:
-                    args = inputs.get(timeout=1)
-                except queue.Empty:
-                    continue
-                try:
-                    run_input(*args)
-                except BaseException:
-                    # This should basically never happen, since only KeyboardInterrupt is the only error that can
-                    # bubble out of from handle_input_exception and those wouldn't be raised outside the main thread
-                    pass
-                inputs.task_done()
-
-        for input_id, function_call_id, args, kwargs in container_io_manager.run_inputs_outputs(
-            imp_fun.input_concurrency
-        ):
-            if spawned_workers < imp_fun.input_concurrency:
-                threading.Thread(target=worker_thread, daemon=True).start()
-                spawned_workers += 1
-            inputs.put((input_id, function_call_id, args, kwargs))
-
-        finished.set()
-        inputs.join()
-
-    else:
-        for input_id, function_call_id, args, kwargs in container_io_manager.run_inputs_outputs(
-            imp_fun.input_concurrency
-        ):
-            try:
-                run_input(input_id, function_call_id, args, kwargs)
-            except:
-                raise
-
-
-async def call_function_async(
+def call_function(
+    user_code_event_loop: UserCodeEventLoop,
     container_io_manager,  #: ContainerIOManager,  TODO: this type is generated at runtime
     imp_fun: ImportedFunction,
 ):
-    async def run_input(input_id: str, function_call_id: str, args: Any, kwargs: Any) -> None:
+    async def run_input_async(input_id: str, function_call_id: str, args: Any, kwargs: Any) -> None:
```

Comment:
Do you think we could move out `run_input_async` and `run_input_sync` into their own top-level functions now that they've gotten bigger? They could take `container_io_manager` and `imp_fun` as parameters.

Reply:
Yeah that makes sense